src/main/kotlin/no/skatteetaten/aurora/boober/feature/HerkimerVaultFeature.kt

```diff
@@ -0,0 +1,206 @@
+
+        if (validationContext) {
+            return resources
+        }
+
+        val herkimerResources: Map<HerkimerVaultResource, List<ResourceHerkimer>> =
+            herkimerVaultResources.associateWith {
+                herkimerService.getClaimedResources(spec.applicationDeploymentId, it.serviceClass)
+            }
+
+        val secrets: Map<HerkimerVaultResource, List<Secret>> = herkimerResources.map { (config, resources) ->
+            val secrets = if (!config.multiple) {
+                listOf(generateHerkimerSecret(resources.first(), spec))
+            } else {
+                resources.map {
+                    generateHerkimerSecret(it, spec)
+                }
+            }
+            config to secrets
+        }.toMap()
+
+        return resources
+            .addIfNotNull(HERKIMER_REPONSE_KEY to herkimerResources)
+            .addIfNotNull(HERKIMER_SECRETS_KEY to secrets)
+    }
+
+    override fun validate(
+        adc: AuroraDeploymentSpec,
+        fullValidation: Boolean,
+        context: FeatureContext
+    ): List<Exception> {
+
+        val configuredHerkimerVaultResources = context.configuredHerkimerVaultResources
+
+        // TODO: Multiple respones with same key and more then 1 are single
+        val errors = validateNotExistsResourcesWithMultipleTrueAndFalseWithSamePrefix(configuredHerkimerVaultResources)
+
+        if (!fullValidation) return errors
+
+        val herkimerResponses = context.herkimerResponse
+
+        val fullValidationErrors = herkimerResponses.filter { !it.key.multiple && it.value.size > 1 }.map {
+            IllegalStateException("Resource with key=${it.key.key} expects a single result but ${it.value.size} was returned")
+        }
+        return errors.addIfNotNull(fullValidationErrors)
+    }
+
+    override fun generate(adc: AuroraDeploymentSpec, context: FeatureContext): Set<AuroraResource> {
+        return context.herkimerVaultResources.values.flatMap { secrets ->
+            secrets.map {
+                it.generateAuroraResource()
+            }
+        }.toSet()
+    }
+
+    private fun generateHerkimerSecret(response: ResourceHerkimer, adc: AuroraDeploymentSpec): Secret {
+
+        // TODO: Should this be first and should it be called credentials?
+        // TODO: What do we do if there are multiple claims?
+        val values = jsonMapper().convertValue<Map<String, String>>(response.claims.first().credentials)
+        return newSecret {
+            metadata {
+                name = response.name.ensureStartWith(adc.name, "-")
+                    .ensureEndsWith(HERKIMER_RESOURCE_KEY, "-")
+                    .toLowerCase().replace("_", "-")
+                namespace = adc.namespace
+            }
+            data = values.mapValues { content -> Base64.encodeBase64String(content.value.toByteArray()) }
+        }
+    }
+
+    override fun modify(adc: AuroraDeploymentSpec, resources: Set<AuroraResource>, context: FeatureContext) {
```

Comment:
Dette er et litt kinkig sted. Her foregår mye av "kjernefunksjonaliteten" direkte i en overridden metode som heter modify. Her må man lese og forstå all koden for å skjønne på hvilke måte ting blir modifyet. Jeg ville foreslå at noe trekkes ut en egen metode som er navngitt i forhold til hva som skjer. F.eks.

```
override fun modify(adc: AuroraDeploymentSpec, resources: Set<AuroraResource>, context: FeatureContext) {
    val envVars: List<EnvVar> = context.herkimerVaultResources
        .flatMap { (config, secrets) -> createSecretRefEnvVars(secrets, config.prefix, config.multiple) }

    resources.addEnvVarsToMainContainers(envVars, javaClass)
}

private fun createSecretRefEnvVars(secrets: List<Secret>, prefix: String, multiple: Boolean): List<EnvVar> {
    return (if (multiple) {
        secrets.mapIndexed { index, secret ->
            secret.createEnvVarRefs(prefix = "${prefix}_${index}_")
        }
    } else {
        listOf(secrets.first().createEnvVarRefs(prefix = "${prefix}_"))
    }).flatten()
}
```

Det kan være det vil være lurt å unne på seg en kommentar akkurat her om hva som er strategien for generering av miljøvariabelnavn, dersom noen skal prøve sette seg inn i dette uten å ha klart for seg detaljene først (snakket med noen).

Reply:
Jeg har endret metoden litt, men ikke slik du foreslår. Du kan jo se hva du synes om mitt nye forslag.